blocktx/peer_handler.go

```diff
@@ -365,39 +353,39 @@ func (bs *PeerHandler) insertBlock(blockHash *chainhash.Hash, merkleRoot *chainh
 	return bs.store.InsertBlock(context.Background(), block)
 }
 
-func (bs *PeerHandler) markTransactionsAsMined(blockId uint64, transactionHashes []*chainhash.Hash, merkleTree []string) error {
+func (bs *PeerHandler) markTransactionsAsMined(blockId uint64, merkleTree []*chainhash.Hash, blockHeight uint64) error {
 	start := gocore.CurrentNanos()
 	defer func() {
 		gocore.NewStat("blocktx").NewStat("HandleBlock").NewStat("markTransactionsAsMined").AddTime(start)
 	}()
 
 	txs := make([]*blocktx_api.TransactionAndSource, 0, TransactionStoringInterval)
 	merklePaths := make([]string, 0, TransactionStoringInterval)
+	leaves := merkleTree[:(len(merkleTree)+1)/2]
 
-	for txIndex, hash := range transactionHashes {
+	for txIndex, hash := range leaves {
+		// Everything to the right of the first nil will also be nil, as this is just padding upto the next PoT.
+		if hash == nil {
+			break
+		}
+
+		// Otherwise they're txids, which should have merkle paths calculated.
 		txs = append(txs, &blocktx_api.TransactionAndSource{
 			Hash: hash[:],
 		})
 
-		merklePath := bc.GetTxMerklePath(txIndex, merkleTree)
-
-		merklePathBinary, err := merklePath.String()
+		bump, err := bc.NewBUMPFromMerkleTreeAndIndex(blockHeight, merkleTree, uint64(txIndex))
 		if err != nil {
 			return err
 		}
 
-		// verify merkle path correctness
-		root, err := merklePath.CalculateRoot(hash.String())
+		bumpHex, err := bump.String()
 		if err != nil {
 			return err
 		}
 
-		if root != merkleTree[len(merkleTree)-1] {
-			return errors.New("merkle path calculated produced different root")
-		}
-
-		merklePaths = append(merklePaths, merklePathBinary)
-		if txIndex%TransactionStoringInterval == 0 || txIndex == len(transactionHashes)-1 {
+		merklePaths = append(merklePaths, bumpHex)
+		if txIndex%TransactionStoringInterval == 0 || txIndex == len(leaves)-1 {
```

Comment:
@shotasilagadzetaal @boecklim can you please help clarify this line?

Reply:
Noting that it's not my idea. I just renamed the variable.